(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[90],{3905:function(e,t,n){"use strict";n.d(t,{Zo:function(){return u},kt:function(){return f}});var a=n(7294);function r(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function l(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?l(Object(n),!0).forEach((function(t){r(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):l(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,a,r=function(e,t){if(null==e)return{};var n,a,r={},l=Object.keys(e);for(a=0;a<l.length;a++)n=l[a],t.indexOf(n)>=0||(r[n]=e[n]);return r}(e,t);if(Object.getOwnPropertySymbols){var l=Object.getOwnPropertySymbols(e);for(a=0;a<l.length;a++)n=l[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(r[n]=e[n])}return r}var i=a.createContext({}),c=function(e){var t=a.useContext(i),n=t;return e&&(n="function"==typeof e?e(t):o(o({},t),e)),n},u=function(e){var t=c(e.components);return a.createElement(i.Provider,{value:t},e.children)},p={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},m=a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,l=e.originalType,i=e.parentName,u=s(e,["components","mdxType","originalType","parentName"]),m=c(n),f=r,d=m["".concat(i,".").concat(f)]||m[f]||p[f]||l;return n?a.createElement(d,o(o({ref:t},u),{},{components:n})):a.createElement(d,o({ref:t},u))}));function f(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var l=n.length,o=new Array(l);o[0]=m;var s={};for(var i in t)hasOwnProperty.call(t,i)&&(s[i]=t[i]);s.originalType=e,s.mdxType="string"==typeof e?e:r,o[1]=s;for(var c=2;c<l;c++)o[c]=n[c];return a.createElement.apply(null,o)}return a.createElement.apply(null,n)}m.displayName="MDXCreateElement"},5901:function(e,t,n){"use strict";n.r(t),n.d(t,{frontMatter:function(){return s},metadata:function(){return i},toc:function(){return c},default:function(){return p}});var a=n(2122),r=n(9756),l=(n(7294),n(3905)),o=["components"],s={id:"germline",title:"Germline WGS"},i={unversionedId:"tutorials/germline",id:"tutorials/germline",isDocsHomePage:!1,title:"Germline WGS",description:"This case study looks at whole-genome germline variant calling in a single individual (NA12878).",source:"@site/docs/tutorials/germline.md",sourceDirName:"tutorials",slug:"/tutorials/germline",permalink:"/docs/tutorials/germline",editUrl:"https://github.com/${organizationName}/${projectName}/edit/${branch}/website/docs/tutorials/germline.md",version:"current",frontMatter:{id:"germline",title:"Germline WGS"},sidebar:"docs",previous:{title:"Config files",permalink:"/docs/guides/advanced/configs"}},c=[{value:"Prerequisites",id:"prerequisites",children:[]},{value:"Download data files",id:"download-data-files",children:[]},{value:"Map reads to reference genome",id:"map-reads-to-reference-genome",children:[]},{value:"Call variants",id:"call-variants",children:[]},{value:"Evaluate variant calls",id:"evaluate-variant-calls",children:[]}],u={toc:c};function p(e){var t=e.components,n=(0,r.Z)(e,o);return(0,l.kt)("wrapper",(0,a.Z)({},u,n,{components:t,mdxType:"MDXLayout"}),(0,l.kt)("p",null,"This case study looks at whole-genome germline variant calling in a single individual (NA12878)."),(0,l.kt)("h3",{id:"prerequisites"},"Prerequisites"),(0,l.kt)("ul",null,(0,l.kt)("li",{parentName:"ul"},(0,l.kt)("a",{parentName:"li",href:"https://github.com/samtools/samtools"},"samtools")),(0,l.kt)("li",{parentName:"ul"},(0,l.kt)("a",{parentName:"li",href:"https://github.com/samtools/bcftools"},"bcftools")),(0,l.kt)("li",{parentName:"ul"},(0,l.kt)("a",{parentName:"li",href:"https://github.com/lh3/bwa"},"BWA")),(0,l.kt)("li",{parentName:"ul"},(0,l.kt)("a",{parentName:"li",href:"https://github.com/RealTimeGenomics/rtg-tools"},"RTG Tools")),(0,l.kt)("li",{parentName:"ul"},(0,l.kt)("a",{parentName:"li",href:"https://github.com/luntergroup/octopus"},"Octopus")," (with ",(0,l.kt)("a",{parentName:"li",href:"https://github.com/luntergroup/octopus/wiki/Variant-filtering:-Random-Forest"},"random forests")," installed).")),(0,l.kt)("h3",{id:"download-data-files"},"Download data files"),(0,l.kt)("p",null,"First download raw reads from the Illumina platinum genomes project for individual NA12878:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"$ mkdir ~/fastq && cd ~/fastq\n$ wget https://storage.googleapis.com/genomics-public-data/platinum-genomes/fastq/ERR194147_1.fastq.gz\n$ wget https://storage.googleapis.com/genomics-public-data/platinum-genomes/fastq/ERR194147_2.fastq.gz\n")),(0,l.kt)("p",null,"Next download a copy of the human reference sequence. In this example we use GRCh37 plus a decoy contig (recommended). If you prefer to use GRCh38, be sure to get a copy without alternative contigs or patches (but with a decoy contig), such as the one available ",(0,l.kt)("a",{parentName:"p",href:"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz"},"here"),"."),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"$ mkdir ~/reference && cd ~/reference\n$ wget ftp://ftp-$ trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz && gzip -d hs37d5.fa.gz\n")),(0,l.kt)("p",null,"To evaluate our calls we need a truth set. We use the Genome In a Bottle (GIAB) version 3.3.2 high confidence calls for NA12878 (HG001):"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"$ mkdir ~/vcf/giab && cd ~/vcf/giab\n$ wget ftp://ftp-trace.ncbi.nlm.nih.gov//giab/ftp/release/NA12878_HG001/NISTv3.3.2/GRCh37\n")),(0,l.kt)("h3",{id:"map-reads-to-reference-genome"},"Map reads to reference genome"),(0,l.kt)("p",null,"First we need to index the reference sequence:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"$ cd ~/reference\n$ samtools faidx hs37d5.fa\n$ bwa index hs37d5.fa\n")),(0,l.kt)("p",null,"Then map our reads to the reference:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},'$ mkdir ~/bam \n$ bwa mem -t 15 -R "@RG\\tID:1\\tSM:NA12878\\tLB:platinum\\tPU:illumina" \\\n      ~/reference/hs37d5.fa \\\n      ~/fastq/ERR194147_1.fastq.gz ~/fastq/ERR194147_2.fastq.gz \\\n      | samtools view -bh | samtools sort -@ 5 -o ~/bam/NA12878.platinum.b37.bam -\n$ samtools index ~/bam/NA12878.platinum.b37.bam\n')),(0,l.kt)("p",null,"This should take a few hours."),(0,l.kt)("h3",{id:"call-variants"},"Call variants"),(0,l.kt)("p",null,"We do not recommend pre-processing the raw BWA alignments (e.g. duplicate marking, or base quality score recalibration) as we do not find this provides consistent  improvements in accuracy, and tends to slow down calling as pre-processed reads files can be larger than the originals. As this is human data, the default arguments for octopus should work well. We restrict calling to the autosomes plus X as these are the only contigs present in the validation sets. We also request a 'legacy' VCF file to use for benchmarking (see section on octopus's default VCF format)."),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"$ octopus \\\n     -R ~/reference/hs37d5.fa \\\n     -I ~/bam/NA12878.platinum.b37.bam \\\n     -T 1 to MT \\\n     --sequence-error-model PCR-FREE.HISEQ-2000 \\\n     --forest forests/germline.v0.6.0-beta.forest \\\n     -o ~/vcf/NA12878.platinum.b37.octopus.vcf.gz \\\n     --threads 20 \\\n     --legacy\n")),(0,l.kt)("p",null,"This should complete in a few hours."),(0,l.kt)("h3",{id:"evaluate-variant-calls"},"Evaluate variant calls"),(0,l.kt)("p",null,"Finally, we will evaluate our calls with RTG Tools ",(0,l.kt)("inlineCode",{parentName:"p"},"vcfeval"),". This command requires the reference sequence to be preprocessed:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"$ ~/tools/rtgtools/rtg format -o ~/reference/hs37d5_sdf ~/reference/hs37d5.fa\n")),(0,l.kt)("p",null,"Then run vcfeval:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"$ rtg vcfeval \\\n     -t ~/reference/hs37d5_sdf \\\n     -b ~/vcf/giab/HG001_GRCh37_truth.vcf.gz \\\n     --evaluation-regions ~/vcf/giab/HG001_GRCh37_hiconf.bed \\\n     -c ~/vcf/NA12878.platinum.b37.octopus.legacy.vcf.gz \\\n     -o ~/eval/NA12878.platinum.b37.octopus.eval \\\n     -f FORMAT.RFQUAL \\\n     --ref-overlap\n")),(0,l.kt)("p",null,"We see the following results:"),(0,l.kt)("pre",null,(0,l.kt)("code",{parentName:"pre",className:"language-shell"},"Threshold  True-pos-baseline  True-pos-call  False-pos  False-neg  Precision  Sensitivity  F-measure\n----------------------------------------------------------------------------------------------------\n    4.770            3680280        3703892       2042      10581     0.9994       0.9971     0.9983\n     None            3681055        3704782       3297       9806     0.9991       0.9973     0.9982\n")))}p.isMDXComponent=!0}}]);